### PR TITLE
cdiff: Inital rust port of the cdiff library: cdiff-apply

### PR DESCRIPTION
**Pull request by micahsnyder**
This is a rebased version of https://github.com/Cisco-Talos/clamav/pull/302 with merge conflicts fixed and with the `eprintln!()` macros replaced with `debug!()`, `error!()`, etc now that the `libclamav_rust::logging` module is merged into `main`.

99% of the work done by @jhumlick, not me. Thanks John!

### CODE REVIEW

File: libclamav_rust/src/cdiff.rs [hunk 1]
```diff
@@ -0,0 +1,1273 @@
...
+
+    let mut header_length: usize = 0;
+
+    // Only read dsig, header, etc. if this is a cdiff file
+    if is_cdiff {
+        let dsig = read_dsig(&mut file);
+        let dsig = match dsig {
+            Ok(dsig) => dsig,
+            Err(e) => {
+                debug!("{:?}", e.to_string());
```
Comment (jhumlick) [inline libclamav_rust/src/cdiff.rs:235]: Fatal errors should not be shown only at the debug log level.

File: libclamav_rust/src/cdiff.rs [hunk 2]
```diff
@@ -0,0 +1,1273 @@
...
+        let versig_result = unsafe {
+            cli_versig2(
+                sha256.to_vec().as_ptr(),
+                dsig_cstring.as_ptr(),
+                PSS_NSTR.as_ptr(),
+                PSS_ESTR.as_ptr(),
+            )
+        };
+        if versig_result != 0 {
+            debug!("cdiff_apply: Incorrect digital signature");
```
Comment (jhumlick) [inline libclamav_rust/src/cdiff.rs:286]: Fatal errors should not be shown only at the debug log level.

File: libclamav_rust/src/cdiff.rs [hunk 3]
```diff
@@ -0,0 +1,1273 @@
...
+        debug!(
+            "Header len is {}, file len is {}, header offset is {}",
+            header_len, file_len, header_offset
+        );
+
+        let current_pos = file.seek(SeekFrom::Start(header_offset as u64));
+        let current_pos = match current_pos {
+            Ok(current_pos) => current_pos as usize,
+            Err(e) => {
+                debug!("{}", e.to_string());
```
Comment (jhumlick) [inline libclamav_rust/src/cdiff.rs:308]: Fatal errors should not be shown only at the debug log level.

File: libclamav_rust/src/cdiff.rs [hunk 4]
```diff
@@ -0,0 +1,1273 @@
...
+        open_db: None,
+        add_start: None,
+        del_start: None,
+        xchg_start: None,
+    };
+
+    match process_lines(&mut ctx, reader, header_length) {
+        Ok(_) => 0,
+        Err(e) => {
+            debug!("{}", e.to_string());
```
Comment (jhumlick) [inline libclamav_rust/src/cdiff.rs:335]: Fatal errors should not be shown only at the debug log level.

**Review by jhumlick — CHANGES_REQUESTED** [review]
debug! should not be used in cases where a fatal error occurs.

**Review by jhumlick — APPROVED** [review]

**Review by jhumlick — APPROVED** [review]

**Review by jhumlick — APPROVED** [review]

**Comment by micahsnyder** [comment]
Looks like we also have a Windows build failure and a CIFuzz build failure.  I'll try to figure out what's going on with those when I'm able to look at it some more.

**Comment by micahsnyder** [comment]
Replaced by https://github.com/Cisco-Talos/clamav/pull/400

**Pull request closed without merging**